core/store/src/trie/outgoing_metadata.rs

```diff
@@ -0,0 +1,451 @@
+    pub fn load(
+        trie: &dyn TrieAccess,
+        shard_ids: impl IntoIterator<Item = ShardId>,
+        groups_config: ReceiptGroupsConfig,
+        protocol_version: ProtocolVersion,
+    ) -> Result<Self, StorageError> {
+        if !ProtocolFeature::BandwidthScheduler.enabled(protocol_version) {
+            return Ok(Self::new(groups_config));
+        }
+
+        let mut metadatas = BTreeMap::new();
+        for shard_id in shard_ids.into_iter() {
+            let metadata = ReceiptGroupsQueue::load(trie, shard_id, groups_config)?;
+            if let Some(metadata) = metadata {
+                metadatas.insert(shard_id, metadata);
+            }
+        }
+        Ok(Self { metadatas, groups_config })
+    }
+
+    /// Update the metadata when a new receipt is added at the end of the outgoing receipts buffer.
+    /// If the metadata doesn't exist for the shard, a new one will be created.
+    pub fn update_on_receipt_pushed(
+        &mut self,
+        shard_id: ShardId,
+        receipt_size: ByteSize,
+        receipt_gas: Gas,
+        state_update: &mut TrieUpdate,
+    ) -> Result<(), StorageError> {
+        let metadata = self
+            .metadatas
+            .entry(shard_id)
+            .or_insert_with(|| ReceiptGroupsQueue::new(shard_id, self.groups_config));
+        metadata.update_on_receipt_pushed(receipt_size, receipt_gas, state_update)
+    }
+
+    /// Update the metadata when a receipt is removed from the front of the outgoing receipts buffer.
+    /// Can be called only for receipts for which `update_on_receipt_pushed` was called before.
+    pub fn update_on_receipt_popped(
+        &mut self,
+        shard_id: ShardId,
+        receipt_size: ByteSize,
+        receipt_gas: Gas,
+        state_update: &mut TrieUpdate,
+    ) -> Result<(), StorageError> {
+        let metadata = self
+            .metadatas
+            .get_mut(&shard_id)
+            .expect("Metadata for this shard should've been created when a receipt was pushed.");
+        metadata.update_on_receipt_popped(receipt_size, receipt_gas, state_update)
+    }
+
+    /// Get metadata for the outgoing buffer to this shard.
+    pub fn get_metadata_for_shard(&self, shard_id: &ShardId) -> Option<&ReceiptGroupsQueue> {
+        self.metadatas.get(shard_id)
+    }
+}
+
+/// Information about a group of consecutive receipts stored in the outgoing buffer.
+#[derive(Debug, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub enum ReceiptGroup {
+    V0(ReceiptGroupV0),
+}
+
+#[derive(Debug, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub struct ReceiptGroupV0 {
+    /// Total size of receipts in this group.
+    /// Should be no larger than `max_receipt_size`, otherwise the bandwidth
+    /// scheduler will not be able to grant the bandwidth neeed to send
+    /// the receipts in this group.
+    pub size: u64,
+    /// Total gas of receipts in this group.
+    pub gas: u128,
+}
```

Comment:
Yeah it's the same as in congestion info.
`ByteSize` is not borsh serializable, so it can't be stored and I used u64 instead.
`Gas` is `u64`, which could potentially overflow (300 TGAS is 300e12,  3000 receipts * 300 TGAS = 1e18), so I used u128 to store the accumullated gas.

Reply:
I didn't put the queue sizes in `CongestionInfo` because then it'd be included in the chunk header, increasing its size.

Reply:
In congestion info we also have u128 for gas and u64 for size.